File: java/src/test/java/com/gergo/takacs/linearAlgebra/VectorTest.java
package com.gergo.takacs.linearAlgebra;

import org.junit.Test;

import java.util.Collections;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class VectorTest {
    @Test(expected = EmptyCollectionNotPermitted.class)
    public void testConstructor() throws Exception {
        //given
        //when
        new Vector(Collections.emptyList());
        //then
    }

    @Test
    public void testAdd() throws Exception {
        //given
        Vector expected = new Vector(2d, 3d, 4d);
        Vector vector1 = new Vector(1d, 2d, 3d);
        Vector vector2 = new Vector(1d, 1d, 1d);
        //when
        Vector actual = vector1.add(vector2);
        //then
        assertEquals(expected, actual);
    }

    @Test
    public void testSub() throws Exception {
        //given
        Vector expected = new Vector(2d, 3d, 4d);
        Vector vector1 = new Vector(3d, 4d, 5d);
        Vector vector2 = new Vector(1d, 1d, 1d);
        //when
        Vector actual = vector1.sub(vector2);
        //then
        assertEquals(expected, actual);
    }

    @Test
    public void testScalarProduct() throws Exception {
        //given
        Vector expected = new Vector(2d, 3d, 4d);
        Vector vector = new Vector(1d, 1.5, 2d);
        //when
        Vector actual = vector.scalarProduct(2);
        //then
        assertEquals(expected, actual);
    }

    @Test
    public void testDotProduct() throws Exception {
        //given
        double expected = 18;
        Vector vector1 = new Vector(2d, 3d);
        Vector vector2 = new Vector(3d, 4d);
        //when
        double actual = vector1.dotProduct(vector2);
        //then
        assertEquals(expected, actual, 10e-10);
    }

    @Test
    public void testMagnitude() throws Exception {
        //given
        double expected = 5;
        Vector vector = new Vector(3d, 4d);
        //when
        double actual = vector.magnitude();
        //then
        assertEquals(expected, actual, 10e-10);
    }

    @Test
    public void testNormalize() throws Exception {
        //given
        Vector expected = new Vector(3 / 5., 4 / 5.);
        Vector vector = new Vector(3d, 4d);
        //when
        Vector actual = vector.normalize();
        //then
        assertEquals(expected, actual);
    }

    @Test
    public void testAngle() throws Exception {
        //given
        double expected = Math.PI / 2;
        Vector vector1 = new Vector(3d, 4d);
        Vector vector2 = new Vector(4d, -3d);
        //when
        double actual = vector1.angle(vector2);
        //then
        assertEquals(expected, actual, 10e-10);
    }

    @Test
    public void testAngleDeg() throws Exception {
        //given
        double expected = 90;
        Vector vector1 = new Vector(3d, 4d);
        Vector vector2 = new Vector(4d, -3d);
        //when
        double actual = vector1.angleDeg(vector2);
        //then
        assertEquals(expected, actual, 10e-10);
    }

    @Test
    public void testIsOrthogonal() throws Exception {
        //given
        Vector vector1 = new Vector(3d, 4d);
        Vector vector2 = new Vector(4d, -3d);
        //when
        boolean actual = vector1.isOrthogonal(vector2);
        //then
        assertTrue(actual);
    }

    @Test
    public void testIsParallelIfReverse() throws Exception {
        //given
        Vector vector1 = new Vector(3d, 4d);
        Vector vector2 = new Vector(-3d, -4d);
        //when
        boolean actual = vector1.isParallel(vector2);
        //then
        assertTrue(actual);
    }

    @Test
    public void testIsParallelIfSameDir() throws Exception {
        //given
        Vector vector1 = new Vector(3d, 4d);
        Vector vector2 = new Vector(6d, 8d);
        //when
        boolean actual = vector1.isParallel(vector2);
        //then
        assertTrue(actual);
    }

    @Test
    public void testProjectTo() throws Exception {
        //given
        Vector expected = new Vector(2, 0d);
        Vector vector1 = new Vector(2d, 20d);
        Vector vector2 = new Vector(3d, 0d);
        //when
        Vector actual = vector1.projectTo(vector2);
        //then
        assertEquals(expected, actual);
    }

    @Test
    public void testProjectToOrthogonal() throws Exception {
        //given
        Vector expected = new Vector(0, 20d);
        Vector vector1 = new Vector(2d, 20d);
        Vector vector2 = new Vector(3d, 0d);
        //when
        Vector actual = vector1.projectToOrthogonal(vector2);
        //then
        assertEquals(expected, actual);
    }

    @Test
    public void testCrossProductIf3D() throws Exception {
        //given
        Vector expected = new Vector(0, 0d, 20d);
        Vector vector1 = new Vector(5d, 0d, 0d);
        Vector vector2 = new Vector(0, 4d, 0d);
        //when
        Vector actual = vector1.crossProduct(vector2);
        //then
        assertEquals(expected, actual);
    }

    @Test
    public void testCrossProductIf2D() throws Exception {
        //given
        Vector expected = new Vector(0, 0d, 20d);
        Vector vector1 = new Vector(5d, 0d);
        Vector vector2 = new Vector(0, 4d);
        //when
        Vector actual = vector1.crossProduct(vector2);
        //then
        assertEquals(expected, actual);
    }

    @Test
    public void testParallelogramArea() throws Exception {
        //given
        double expected = 20d;
        Vector vector1 = new Vector(5d, 0d);
        Vector vector2 = new Vector(0, 4d);
        //when
        double actual = vector1.parallelogramArea(vector2);
        //then
        assertEquals(expected, actual, 10e-10);
    }

    @Test
    public void testTriangleArea() throws Exception {
        //given
        double expected = 10d;
        Vector vector1 = new Vector(5d, 0d);
        Vector vector2 = new Vector(0, 4d);
        //when
        double actual = vector1.triangleArea(vector2);
        //then
        assertEquals(expected, actual, 10e-10);
    }


}

File: scala/src/test/scala/com/gergo/takacs/linearAlgebra/VectorTest.scala
package com.gergo.takacs.linearAlgebra

import org.scalatest.{FlatSpec, Matchers}


class VectorTest extends FlatSpec with Matchers {

  "Vector dimension" should "be equal with the length of the creator collection" in {
    val underTest = Vector(2, 2, 3)
    underTest.dimension() should be(3)
  }

  it should "throw exception in case of empty Collection" in {
    a[EmptyCollectionNotPermitted] should be thrownBy {
      Vector(List())
    }
  }

  "Add vector to this" should "create a new vector with coordinates added each" in {
    val actual = Vector(2, 3) + Vector(2, 3)
    actual should be(Vector(4, 6))
  }

  "Subtract vector to this" should "create a new vector with coordinates subtracted each" in {
    val actual = Vector(6, 3) - Vector(2, 3)
    actual should be(Vector(4, 0))
  }

  "Vector multiply by scalar" should "have scalar times coordinates" in {
    val actual = Vector(3, 4) * 2
    actual should be(Vector(6, 8))
  }

  "Scalar multiply by Vector" should "have scalar times coordinates" in {
    val actual = 2 *: Vector(3, 4)
    actual should be(Vector(6, 8))
  }

  "Vector multiply by Vector" should "return with inner product" in {
    val actual = Vector(2, 3) * Vector(3, 4)
    actual should be(18)
  }

  "Vector magnitude" should "return with vector length" in {
    val actual = Vector(3, 4).magnitude()
    actual should be(5)
  }

  "Vector normalization" should "return a vector with length of 1" in {
    val actual = Vector(3, 4).normalize()
    actual should be(Vector(.6, .8))
  }

  "Angle of two vectors" should "return a radian value" in {
    val actual = Vector(2, 3).angle(Vector(3, -2))
    actual should be(math.Pi / 2)
  }

  "Angle of two vectors in degrees" should "return a radian value" in {
    val actual = Vector(2, 3).angleDeg(Vector(3, -2))
    actual should be(90)
  }

  "If two Vector is orthogonal" should "return true" in {
    val actual = Vector(2, 3).isOrthogonal(Vector(3, -2))
    actual should be(true)
  }

  "If two Vector is parallel" should "return true" in {
    val actual = Vector(2, 3).isParallel(Vector(6, 9))
    actual should be(true)
  }

  "If two Vector is parallel but different direction" should "return true" in {
    val actual = Vector(2, 3).isParallel(Vector(-6, -9))
    actual should be(true)
  }

  "If null vector isZero" should "return true" in {
    val actual = Vector(0, 0).isZero()
    actual should be(true)
  }

  "Project a vector to an other one" should "return a parallel vector with the base" in {
    val actual = Vector(2, 20).projectTo(Vector(3, 0))
    actual should be(Vector(2, 0))
  }

  "Project a vector to the orthogonal of an other vector" should "return a orthogonal vector" in {
    val actual = Vector(2, 20).projectToOrthogonal(Vector(3, 0))
    actual should be(Vector(0, 20))
  }

  "Two Vector cross product" should "return a vector" in {
    val actual = Vector(5, 0, 0).crossProduct(Vector(0, 4, 0))
    actual should be(Vector(0, 0, 20))
  }

  "Two Vector cross product for 2D" should "return a vector" in {
    val actual = Vector(5, 0).crossProduct(Vector(0, 4))
    actual should be(Vector(0, 0, 20))
  }

  "Two Vector parallelogram area" should "be the length of cross product vector" in {
    val actual = Vector(5, 0, 0).parallelogramArea(Vector(0, 4, 0))
    actual should be(20)
  }

  "Two Vector triangle area" should "be half of the length of cross product vector" in {
    val actual = Vector(5, 0, 0).triangleArea(Vector(0, 4, 0))
    actual should be(10)
  }
}
